Public/Src/Tools/Execution.Analyzer/Analyzers.Core/XLGPlusPlus/XldbProtobufExtensions.cs

```diff
@@ -558,5 +548,479 @@ public static Xldb.Fingerprint ToFingerprint(this Fingerprint fingerprint)
+                PageStreamsCount = pipTable.PageStreamsCount,
+                Size = pipTable.Size,
+                WritesMilliseconds = pipTable.WritesMilliseconds,
+                ReadsMilliseconds = pipTable.ReadsMilliseconds,
+                Used = pipTable.Used,
+                Alive = pipTable.Alive
+            };
+
+            xldbPipTable.StableKeys.AddRange(pipTable.StableKeys.Select(stableKey => stableKey.Value));
+            xldbPipTable.Keys.AddRange(pipTable.Keys.Select(key => key.Value));
+            xldbPipTable.DeserializationContexts.AddRange(pipTable.DeserializationContexts.Select(
+                context => new PipDeserializationContext() { Key = (PipQueryContext)context.Key, Value = context.Value }));
+            return xldbPipTable;
+        }
+
+        /// <nodoc />
+        public static Xldb.PipData ToPipData(this PipData pipData)
+        {
+            return !pipData.IsValid ? null : new Xldb.PipData
+            {
+                IsValid = pipData.IsValid,
+                FragmentSeparator = pipData.FragmentSeparator.ToString(),
+                FragmentCount = pipData.FragmentCount,
+                FragmentEscaping = (PipDataFragmentEscaping)pipData.FragmentEscaping
+            };
+
+        }
+
+        /// <nodoc />
+        public static Xldb.PipProvenance ToPipProvenance(this PipProvenance provenance)
+        {
+            return provenance == null ? null : new Xldb.PipProvenance()
+            {
+                Usage = provenance.Usage.ToPipData(),
+                ModuleId = provenance.ModuleId.Value.Value,
+                ModuleName = provenance.ModuleName.ToString(),
+                SemiStableHash = provenance.SemiStableHash
+            };
+        }
+
+        public static Xldb.FileOrDirectoryArtifact ToFileOrDirectoryArtifact(this FileOrDirectoryArtifact artifact, PathTable pathTable)
+        {
+            var xldbFileOrDirectoryArtifact = new Xldb.FileOrDirectoryArtifact()
+            {
+                IsValid = artifact.IsValid
+            };
+
+            if (artifact.IsDirectory)
+            {
+                xldbFileOrDirectoryArtifact.IsDirectory = true;
+                xldbFileOrDirectoryArtifact.DirectoryArtifact = artifact.DirectoryArtifact.ToDirectoryArtifact(pathTable);
+            }
+            else
+            {
+                xldbFileOrDirectoryArtifact.IsFile = true;
+                xldbFileOrDirectoryArtifact.FileArtifact = artifact.FileArtifact.ToFileArtifact(pathTable);
+            }
+
+            return xldbFileOrDirectoryArtifact;
+        }
+
+        /// <nodoc />
+        public static Xldb.Pip ToPip(this Pip pip, DirectedGraph directedGraph)
+        {
+            var xldbPip = new Xldb.Pip()
+            {
+                SemiStableHash = pip.SemiStableHash,
+                FormattedSemiStableHash = pip.FormattedSemiStableHash,
+                ProcessAllowsUndeclaredSourceReads = pip.ProcessAllowsUndeclaredSourceReads,

```

Comment:
>ProcessAllowsUndeclaredSourceReads [](start = 16, length = 34)

This seems very specific to process pips. I wouldn't include it on Xldb.Pip

Reply:
Good point, ill go ahead and remove it. 

---
In reply to: [314445602](https://github.com/microsoft/BuildXL/pull/724#discussion_r314445602) [](ancestors = 314445602)